server/src/main/scala/manifests/graph/BatchStage.scala

```diff
@@ -58,43 +85,54 @@ class BatchStage(now: () => SDateLike) extends GraphStage[FlowShape[List[Arrival
       }
     }
 
-    private def prioritiseAndPush(existingSubscribers: Map[ArrivalKey, Option[MillisSinceEpoch]], existingPrioritised: Set[ArrivalKey]): Unit = {
-      val prioritisedBatch = updatePrioritisedAndSubscribers(existingSubscribers, existingPrioritised)
+    private def prioritiseAndPush(): Unit = {
+      purgeExpired()
+      val lookupBatch = updatePrioritisedAndSubscribers()
 
-      if (prioritisedBatch.nonEmpty) {
-        log.info(s"Pushing ${prioritisedBatch.size} prioritised arrivals. ${lookupQueue.size} prioritised remaining.")
-        push(outArrivals, prioritisedBatch.toList)
+      if (lookupBatch.nonEmpty) {
+        log.info(s"Pushing ${lookupBatch.size} lookup requests. ${lookupQueue.size} lookup requests remaining.")
+        push(outArrivals, lookupBatch.toList)
       } else log.info(s"Nothing to push right now")
     }
 
-    private def updatePrioritisedAndSubscribers(existingSubscribers: Map[ArrivalKey, Option[MillisSinceEpoch]], existingPrioritised: Set[ArrivalKey]): Set[ArrivalKey] = {
-      log.info(s"about to check all arrivals for those due a lookup")
-      val updatedPrioritised: Set[ArrivalKey] = addToPrioritised(existingSubscribers, existingPrioritised)
+    private def pushRegisteredArrivalsUpdates(): Unit = if (registeredArrivalsUpdates.nonEmpty) {
+      log.info(s"Pushing ${registeredArrivalsUpdates.size} registered arrivals updates")
+      push(outRegisteredArrivals, RegisteredArrivals(registeredArrivalsUpdates))
+      registeredArrivalsUpdates = Map()
+    }
+
+    private def purgeExpired(): Unit = {
+      lookupQueue = Crunch.purgeExpired(lookupQueue, (a: ArrivalKey) => a.scheduled, now, expireAfterMillis)
+      registeredArrivals = Crunch.purgeExpiredTuple(registeredArrivals, (a: ArrivalKey) => a.scheduled, now, expireAfterMillis)
+    }
 
-      val (prioritisedBatch, remainingPrioritised) = updatedPrioritised.splitAt(500)
+    private def updatePrioritisedAndSubscribers(): Set[ArrivalKey] = {
+      val currentNow = now()
```

Comment:
this could maybe do with a type annotation - is it a millis or an SDate?

Reply:
done